# Project Description - What Your Shoes Say About You?
Can we predict other people’s personalities, just by taking a photo of their shoes? What about yourself, - Are you a go-getter, an attention-seeker, or a loyal friend? You can try this model on yourself to see if it predicts your personality correctly. Resources: This is an open-ended problem, and you may conduct your own research and find relevant papers or datasets. Here are some suggestions, but you don’t have to use these.
This article explains how the type and shape of shoes are related to various personalities:
https://www.rd.com/list/shoes-personality/

In this notebook I have analyzed 6 different classes. To do this we performed the following set of tasks:
1. Import the set of images
2. Building an appropriate CNN model
3. Training the CNN model
4. Making Prediction off the trained model
5. Checking Accuracy of Model

### Import Libraries

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [2]:
import tensorflow as tf
from keras.preprocessing.image import ImageDataGenerator

### Data Preprocessing

In our dataset we are looking at 6 different classes of shoes:
- Ankle high boots
- Knee high boots
- Athletic sandals
- Boat shoes
- Oxford shoes
- Sneakers
<br>

### Preprocessing the Training Set


In [3]:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)
training_set = train_datagen.flow_from_directory('/content/gdrive/My Drive/Colab Notebooks/MANU465/Dataset/Train',
                                                 target_size = (64, 64),
                                                 batch_size = 32,
                                                 class_mode = 'categorical')

Found 856 images belonging to 6 classes.


### Data Preprocessing

### Preprocessing the Test Set

In [4]:
test_datagen = ImageDataGenerator(rescale = 1./255)
test_set = test_datagen.flow_from_directory('/content/gdrive/My Drive/Colab Notebooks/MANU465/Dataset/Test',
                                            target_size = (64, 64),
                                            batch_size = 32,
                                            class_mode = 'categorical')

Found 201 images belonging to 6 classes.


In [5]:
Model = tf.keras.models.Sequential()

Model.add(tf.keras.layers.Conv2D(filters=3, kernel_size=3, activation='relu', input_shape=[64, 64, 3]))

In [6]:
Model.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))

In [7]:
Model.add(tf.keras.layers.Conv2D(filters=3, kernel_size=3, activation='relu'))

In [8]:
Model.add(tf.keras.layers.MaxPool2D(pool_size=2, strides=2))

In [9]:
Model.add(tf.keras.layers.Flatten())

In [10]:
Model.add(tf.keras.layers.Dense(units=128, activation='relu'))

In [11]:
Model.add(tf.keras.layers.Dense(units=6, activation='softmax')) # units = number of classes

In [12]:
Model.compile(loss='categorical_crossentropy', optimizer='adam',metrics=['accuracy'])

In [13]:
Model.fit(x = training_set, validation_data = test_set, epochs = 25) #check with 20, 25 and 30

Epoch 1/30
27/27 [==============================] - 7s 211ms/step - loss: 1.4651 - accuracy: 0.4030 - val_loss: 1.2749 - val_accuracy: 0.4776
Epoch 2/30
27/27 [==============================] - 6s 233ms/step - loss: 0.9839 - accuracy: 0.6449 - val_loss: 0.9552 - val_accuracy: 0.6368
Epoch 3/30
27/27 [==============================] - 5s 190ms/step - loss: 0.8198 - accuracy: 0.6928 - val_loss: 0.7798 - val_accuracy: 0.6866
Epoch 4/30
27/27 [==============================] - 5s 180ms/step - loss: 0.6952 - accuracy: 0.7535 - val_loss: 0.7955 - val_accuracy: 0.6517
Epoch 5/30
27/27 [==============================] - 5s 186ms/step - loss: 0.6580 - accuracy: 0.7558 - val_loss: 0.7341 - val_accuracy: 0.6567
Epoch 6/30
27/27 [==============================] - 5s 181ms/step - loss: 0.6170 - accuracy: 0.7862 - val_loss: 0.7545 - val_accuracy: 0.6965
Epoch 7/30
27/27 [==============================] - 5s 191ms/step - loss: 0.5820 - accuracy: 0.7839 - val_loss: 0.6706 - val_accuracy: 0.7562
Epoch 

In [14]:
Model.evaluate(training_set)


27/27 [==============================] - 4s 138ms/step - loss: 0.2852 - accuracy: 0.8960


[0.2852030098438263, 0.8960280418395996]

In [15]:
Model.evaluate(test_set)

7/7 [==============================] - 1s 87ms/step - loss: 0.6317 - accuracy: 0.7910


[0.6317297220230103, 0.7910447716712952]

Given the results of our model with 20 epochs the accuracy was ~90% and 75% ofn the test set. This signifies a strong accuracy witht he following model built. Additionally, not included in the final notebook the model was tested with a 25 epochs as well. 
<br><br>
With 25 epochs of the same configuration (loss=cateogrical_crossentropy) and (optimzier=nadam) the results were 0.877 and 0.766. for the accuracy.
<br><br>
Another test was done for a different configuration of (loss=cateogrical_crossentropy) and (optimzier=sgd) and with 25 epochs the results were an accuracy of 0.711 on the test set and 0.8 on the training set 
<br><br>
Another test was done for a different configuration of (loss=cateogrical_crossentropy) and (optimzier=sgd) and with 25 epochs the results were an accuracy of 0.711 on the test set and 0.8 on the training set 
<br><br>
One final test was done with the following configuration:  loss=cateogrical_crossentropy and optimzier=adam and with 25 epochs the results were an accuracy of 0.791 on the test set and 0.91 on the training set 
<br><br>
This points to being our best solution model as it was the most accurate. With this revelation I also decided to try once more with 30 epochs. Interestingly the test set results decreased to 0.77 while the training set accuracy increased to 0.93. Moreover, this is a clear sign of overfitting as the test set results got worse. Therefore the final model we decide to proceed with is the set up described previously.
<br><br>
Moreover for the dataset chosen we can see that the 6 types of shoes/personalities can be predicted well. These being:
1. Sneakers - Running shoe fans are goal-oriented/Sneaker buffs get along with everyone
2. Sandals - Flip-flop fans are laid back
3. Oxfords - Loafer lovers are fastidious
4. Boat shoes - They give an impression of a natural, cool attitude. It represents the style of the voyager. (https://whatisfullformof.com/choose-your-shoes-wisely-they-are-your-identity/)
5.  Ankle High Boots - You’re an Ace Multitasker (https://www.womansworld.com/posts/fashion/fall-boot-style-personality-173323)
6. Knee High Boots - You’re a Bold Leader (https://www.womansworld.com/posts/fashion/fall-boot-style-personality-173323)



